tests/nextflowtower/test_client.py

```diff
@@ -0,0 +1,111 @@
+import json
+import os
+from types import GeneratorType
+
+import pytest
+
+from sagetasks.nextflowtower import client
+
+EG_TOKEN = "token"
+EG_API_URL = "https://example.com"
+EG_METHOD = "GET"
+EG_ENDPOINT = "/example/endpoint"
+
+
+@pytest.fixture
+def tower_client():
+    return client.TowerClient(EG_TOKEN, EG_API_URL)
+
+
+def test_import():
+    assert client
+
+
+class TestTowerClient:
+    def test_missing_args(self):
+        # Clear environment
+        env_vars = [
+            "NXF_TOWER_TOKEN",
+            "TOWER_ACCESS_TOKEN",
+            "NXF_TOWER_API_URL",
+            "TOWER_API_ENDPOINT",
+        ]
+        for var in env_vars:
+            if var in os.environ:
+                del os.environ[var]
+        # Ensure exception when initializing
+        with pytest.raises(KeyError):
+            client.TowerClient()
+
+    def test_get_valid_name(self, tower_client):
```

Comment:
Nit: You can use pytest parametrize to avoid the for loop.  That said, this is simple enough that you don't need to.

Reply:
I tried using `pytest.mark.parametrize()` (see below), but I don't think we gain anything that's worth the added complexity (and messiness) compared to a `for` loop. It also makes this one test take up 7 lines in the `pytest` output, which seems disproportionate for such a simple test. I have used this in other situations though, so I'll keep it in mind. 

## Using a `for` loop

```python
    def test_get_valid_name(self, tower_client):
        expectations = {
            "foobar": "foobar",
            "FooBar": "FooBar",
            "foo-bar": "foo-bar",
            "foo_bar": "foo_bar",
            "f00b4r": "f00b4r",
            "foo,bar": "foo-bar",
            "foo (bar)": "foo--bar-",
        }
        for name, output in expectations.items():
            assert tower_client.get_valid_name(name) == output
```

## Using `pytest.mark.parametrize()`

```python
    @pytest.mark.parametrize(
        "raw,valid",
        [
            ("foobar", "foobar"),
            ("FooBar", "FooBar"),
            ("foo-bar", "foo-bar"),
            ("foo_bar", "foo_bar"),
            ("f00b4r", "f00b4r"),
            ("foo,bar", "foo-bar"),
            ("foo (bar)", "foo--bar-"),
        ],
    )
    def test_get_valid_name(self, tower_client, raw, valid):
        assert tower_client.get_valid_name(raw) == valid
```

Which produces this output:

```console
tests/test_stub.py::test_stub PASSED                                                                                                                        [  7%]
tests/nextflowtower/test_client.py::test_import PASSED                                                                                                      [ 15%]
tests/nextflowtower/test_client.py::TestTowerClient::test_missing_args PASSED                                                                               [ 23%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[foobar-foobar] PASSED                                                              [ 30%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[FooBar-FooBar] PASSED                                                              [ 38%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[foo-bar-foo-bar] PASSED                                                            [ 46%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[foo_bar-foo_bar] PASSED                                                            [ 53%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[f00b4r-f00b4r] PASSED                                                              [ 61%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[foo,bar-foo-bar] PASSED                                                            [ 69%]
tests/nextflowtower/test_client.py::TestTowerClient::test_get_valid_name[foo (bar)-foo--bar-] PASSED                                                        [ 76%]
tests/nextflowtower/test_client.py::TestTowerClient::test_request_nonempty PASSED                                                                           [ 84%]
tests/nextflowtower/test_client.py::TestTowerClient::test_request_empty PASSED                                                                              [ 92%]
tests/nextflowtower/test_client.py::TestTowerClient::test_paged_request PASSED                                                                              [100%]

```